internal/controller/redis_connector_test.go

```diff
@@ -0,0 +1,280 @@
+		account      string
+		nodeID       string
+		hostname     string
+		expectedHost string
+		accIndex     []string
+		connIndex    []string
+		podIndex     []string
+		err          error
+	}{
+		{
+			account:      "01",
+			nodeID:       "node-a",
+			hostname:     testHost,
+			expectedHost: testHost,
+			accIndex:     []string{"node-a:" + testHost},
+			connIndex:    []string{"01:node-a:" + testHost},
+			podIndex:     []string{"01:node-a"},
+			err:          nil,
+		},
+		{
+			account:      "01",
+			nodeID:       "node-a",
+			hostname:     "dupe-conn",
+			expectedHost: testHost,
+			accIndex:     []string{"node-a:" + testHost},
+			connIndex:    []string{"01:node-a:" + testHost},
+			podIndex:     []string{},
+			err:          DuplicateConnectionError{},
+		},
+		{
+			account:      "02",
+			nodeID:       "node-a",
+			hostname:     testHost,
+			expectedHost: testHost,
+			accIndex:     []string{"node-a:" + testHost},
+			connIndex:    []string{"01:node-a:" + testHost, "02:node-a:" + testHost},
+			podIndex:     []string{"01:node-a", "02:node-a"},
+			err:          nil,
+		},
+		{
+			account:      "02",
+			nodeID:       "node-a",
+			hostname:     "dupe-conn",
+			expectedHost: testHost,
+			accIndex:     []string{"node-a:" + testHost},
+			connIndex:    []string{"01:node-a:" + testHost, "02:node-a:" + testHost},
+			podIndex:     []string{},
+			err:          DuplicateConnectionError{},
+		},
+	}
+
+	for _, tc := range tests {
+		got := RegisterWithRedis(c, tc.account, tc.nodeID, tc.hostname)
+		if got != tc.err {
+			t.Fatalf("expected: %v, got: %v", tc.err, got)
+		}
+		assert.Equal(t, c.Get(tc.account+":"+tc.nodeID).Val(), tc.expectedHost)
+		assert.Equal(t, c.SMembers(tc.account).Val(), tc.accIndex)
+		assert.Equal(t, c.SMembers("connections").Val(), tc.connIndex)
+		assert.Equal(t, c.SMembers(tc.hostname).Val(), tc.podIndex)
+	}
+}
+
+func TestUnregisterWithRedis(t *testing.T) {
+	s, _ := miniredis.Run()
+	defer s.Close()
+
+	c := newTestRedisClient(s.Addr())
+
+	_ = RegisterWithRedis(c, "01", "node-a", testHost)
+	_ = RegisterWithRedis(c, "01", "node-b", testHost)
+	_ = RegisterWithRedis(c, "01", "node-c", testHost)
+
+	tests := []struct {
+		account   string
+		nodeID    string
+		remaining []string
+	}{
+		{account: "01", nodeID: "node-a", remaining: []string{"01:node-b", "01:node-c"}},
+		{account: "01", nodeID: "node-not-found", remaining: []string{"01:node-b", "01:node-c"}},
+		{account: "01", nodeID: "node-b", remaining: []string{"01:node-c"}},
+		{account: "01", nodeID: "node-c", remaining: []string{}},
+	}
+
+	for _, tc := range tests {
+		UnregisterWithRedis(c, tc.account, tc.nodeID, testHost)
+		assert.Equal(t, c.Get(tc.account+":"+tc.nodeID).Err(), redis.Nil)
+		assert.Equal(t, c.SMembers(testHost).Val(), tc.remaining)
+	}
+
+	assert.Equal(t, c.SMembers("01").Val(), []string{})
+	assert.Equal(t, c.SMembers("connections").Val(), []string{})
+	assert.Equal(t, c.SMembers(testHost).Val(), []string{})
+}
```

Comment:
This tests makes sure that all the registered connections have been cleaned up.  What about registering a connection that does not get removed?  The test could make sure that the connection that does not get unregistered is not damaged?

Reply:
Sure I can add that in.